java-squid/src/main/java/org/sonar/java/collections/AVLTree.java

```diff
@@ -113,11 +119,53 @@ private void forEach(AVLTree t, PMap.Consumer<K, V> action) {
     if (t.isEmpty()) {
       return;
     }
-    action.accept((K) t.key(), (V) t.value());
     forEach(t.left(), action);
+    action.accept((K) t.key(), (V) t.value());
     forEach(t.right(), action);
   }
 
+  @Override
+  public Iterator<Map.Entry<K, V>> entriesIterator() {
+    return new NodeIterator(this);
+  }
+
+  private class NodeIterator implements Iterator<Map.Entry<K, V>> {
+    private Deque<AVLTree> stack = new ArrayDeque<>();
+
+    public NodeIterator(AVLTree<K, V> node) {
+      descendLeft(node);
+    }
+
+    @Override
+    public boolean hasNext() {
+      return !stack.isEmpty();
+    }
+
+    private void descendLeft(AVLTree node) {
+      if (node != EMPTY) {
+        stack.addLast(node);
+        descendLeft(node.left());
+      }
+    }
+
+    @Override
+    public Map.Entry<K, V> next() {
+      if (!hasNext()) {
+        throw new NoSuchElementException();
+      }
+      AVLTree node = stack.removeLast();
+      if (node.right() != EMPTY) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)